### PR TITLE
VMS: modernise rand_pool_acquire_entropy

### PR DESCRIPTION
**Pull request by levitte**
Stop redefining structures that are already defined in system
headers.  This also means we can stop setting the pointer size
globally, because the system structures will have the correct pointer
sizes either way.  The only exception is passing the right pointer
size to a function.

Stop trying to twist things around with rand(), that's the job of the
DRBG that we feed.

Stop assuming the location of the JPI$_FINALEXC item, look it up
instead.

Signal an exception if the sys$getjpiw call fails (it means the item
list isn't set up right, so works as an assertion, but using VMS
methodology).

Add more items that could serve as entropy source.

### CODE REVIEW

**Comment by levitte** [comment]
Note: together with #6150, the entropy factor will be made larger (i.e. the VMS entropy gatherer will become much more pessimistic about the quality of the entropy sources)

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: "minus one"? What am I missing? It looks like it's just the "number of elements in *that* array". Caller indeed must have space for one extra NULL element, but in the *output* array, not *that*.

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: Formatting nit, too many spaces. [Or too little if intention was to align equal signs in multiple lines.]

Comment (levitte) [reply crypto/rand/rand_vms.c]: It was previously aligned with the assignment of `->ile3$w_length`.  Fixed.

File: crypto/rand/rand_vms.c
Comment (levitte) [reply crypto/rand/rand_vms.c]: Thanks, that's a misplaced comment.  Fixed

**Comment by levitte** [comment]
I currently get a mysterious crash on IA64...  need to resolve that before approval

**Comment by dot-asm** [comment]
Some of output sizes are 8. Does it imply corresponding alignment? If so, it might be the reason. I mean if it's supposed to be aligned, then there are some that won't be. Because number of prior 4-byte sized elements is not even.

**Comment by levitte** [comment]
Ooooh, I think you're on to something.  The data_buffer pointer you're thinking about points at an array of 32-bit integers, and if the 8-byte element happen to be an odd element, it will be positioned across a 64-bit boundary.

I think that the easiest is to make the array a `GENERIC_64` array instead of a `uint32_t` one.  `GENERIC_64` is a union of a 64-bit integer, 2 32-bit integers and 4 16-bit integers, and since we allocate the space for potential 64-bit numbers at every item anyway, we might as well...

**Comment by levitte** [comment]
Or hmmm...  `GENERIC_64` turns out to be a bad idea, that will leave some holes in the array that we send to the pool.  Better then to treat all 8-byte elements separately...

**Comment by levitte** [comment]
That did change something...  now it crashes elsewhere...

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: I think you meant to remove this...

Comment (levitte) [reply crypto/rand/rand_vms.c]: Ah, good catch, thanks

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: If alignment is significant you still need to ensure alignment of data_buffer. For example `union { double align; uint32_t data[OSSL_NELEM... ] buffer;`, and then passing `&buffer.data[total_length]`...

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: But doesn't split means that prepare_item_list would *grow* JPI_items, not overwrite 64-bit queries with 32-bit ones? Or you'd need pair of getjpiw calls...

Comment (levitte) [reply crypto/rand/rand_vms.c]: Yup, I realised that just now.  Fix coming up

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: `&data_buffer[total_length_64bit]` not `&data_buffer[total_length]`

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: On side note, formally you don't need to reassign items_input or items, you can use arguments directly.

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: Wait a sec! total_length is in bytes, so where would &data_buffer[total_length] point? There ought to be your crash. I mean maybe it wasn't about alignment after all...

Comment (dot-asm) [reply crypto/rand/rand_vms.c]: > maybe it wasn't about alignment after all...

But if it is, then data_buffer_64bit still needs alignment *insurance*.

Comment (levitte) [reply crypto/rand/rand_vms.c]: It seems that insurance is still needed

**Comment by levitte** [comment]
Away a few hours, I'll get back to this tonight

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: For consistency change to `sizeof(databuffer[0])`.

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: Is compiler allowed to optimize this as `(8 / ENTROPY_FACTOR) * total_length`? Which would yield 0 for any total_length... In other words I think explicit `(8 * total_length) / ENTROPY_FACTOR` would be appropriate here...

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: "assumed"

File: crypto/rand/rand_vms.c
Comment (levitte) [reply crypto/rand/rand_vms.c]: I thought `*` and `/` were left-to-right operators. Are compilers allowed to move around operations like you suggest, even though that would give different results?

Comment (kaduk) [reply crypto/rand/rand_vms.c]: The language spec mandates an order of operations, which IIRC is left-to-right, but I did not double-check.

Comment (dot-asm) [reply crypto/rand/rand_vms.c]: Note that it was actually a question. But here is another question. How universal is suggestion that * and / are left-to-right? Do you recognize that it actually depends on type?  For example contemporary compilers have no problem optimizing `8*x/4` as `2*x` for signed integer. But not unsigned. [Do you want to know why?] Though older gcc optimizes it as `2*x` even for unsigned. And don't get me started on floating point... Of course `8*x/4` is very different from `8*x/20` and compilers should recognize that. But can you be *sure*? For any of compilers that are out there? And think of it even like this. Does reader of the code have to keep all those details in mind? Actual value of ENTROPY_FACTOR, dependency on type, which type is involved in operation...

Comment (dot-asm) [reply crypto/rand/rand_vms.c]: > every compiler that are out there?

Well, of course one can say that it has to work with just one, VMS... Yeah, consider it a side note...

File: crypto/rand/rand_vms.c
Comment (levitte) [reply crypto/rand/rand_vms.c]: Fixed

File: crypto/rand/rand_vms.c
Comment (levitte) [reply crypto/rand/rand_vms.c]: Thanks for that heads up.  Me, I'm thinking that paranoia is fine, as long as it doesn't cost our sanity.  In this case, I'll leave it be, because the VMS compilers do fine with it, and considering VSI are keeping a close eye on OpenSSL, I wouldn't be surprised if it's one of the test beds for new ones.

**Comment by levitte** [comment]
This is now crash free on VMS.

**Comment by dot-asm** [comment]
I'm confused. At one point you said that alignment insurance [for data_buffer_64bit] is required, but provide none. This doesn't actually contradict assertion that it's crash-free, because compiler is likely to align it, it's custom. Do you want to leave it for somebody else to keep an eye on too? Or maybe alignment is not an issue after all. You can verify this by initializing total_elems_64bit with 1 (and expanding data_buffer_64bit by 1).

File: crypto/rand/rand_vms.c
Comment (dot-asm) [reply crypto/rand/rand_vms.c]: Fair enough. BTW, would a test fail if 0 is passed to rand_pool_add here?

**Comment by levitte** [comment]
I totally spaced on what you might mean with "insurance", sorry.  What do you have in mind?
The experiment with initializing `total_elems_64bit` to 1 resulted in a `%SYSTEM-F-BADPARAM, bad parameter value` on Alpha, but error free on IA64.  This kinda sorta indicates that you're right about alignment not being an actual issue.  I'll investigate (gotta rebuild in debug mode)...

**Comment by dot-asm** [comment]
> "insurance"

https://github.com/openssl/openssl/pull/6151#discussion_r185495817

> %SYSTEM-F-BADPARAM ... you're right about alignment not being an actual issue.

Contrary, this is indication that alignment *is* an issue. Because failure on *one* system is sufficient to classify it as problem.

File: crypto/rand/rand_vms.c
Comment (dot-asm) [reply crypto/rand/rand_vms.c]: The question was if any of the tests executed in equivalent of 'make test' would fail. I mean question is *how* would the keep an eye on it? They obviously won't single-step whole code base in debugger...

**Comment by levitte** [comment]
Re insurance, I separated the buffers in two, one for 64-bit data and one for 32-bit.  Are you telling me that the compiler won't ensure that an individual array - actually any variable - starts at an aligned address?

**Comment by levitte** [comment]
> > %SYSTEM-F-BADPARAM ... you're right about alignment not being an actual issue.
> 
> Contrary, this is indication that alignment *is* an issue. Because failure on *one* system is sufficient to classify it as problem.

What I didn't say was that there is no error on IA64...  that did take me by surprise.  But I may have failed something, so I'm rebuilding from scratch with `total_elems_64bit` initialized to 1.  But do note that I said "kinda sorta".
(wanna know the infuriating thing?  On alpha, when I build in debug mode and run it under the debugger, that error doesn't show up)

**Comment by dot-asm** [comment]
> Re insurance, I separated the buffers in two, one for 64-bit data and one for 32-bit.

And problem there was that amount of preceding 32-bit values could be uneven, so that few 64-bit ones were *guaranteed* to be misaligned.

> Are you telling me that the compiler won't ensure that an individual array - actually any variable - starts at an aligned address?

Compiler is obliged to align at natural type boundary. I mean 32-bit variables/arrays shall be aligned at 32-bit boundaries, 64-bit - at 64-bit. 32-bit variables *can* be aligned at 64-bit boundaries, and as already said, 64-bit compilers customarily align *all* stack objects at 64-bit boundaries. However, they don't have to, hence you can't make assumption that they actually do. Keyword *here* is that both arrays are declared uint32_t, even data_buffer_64bit, hence compiler *may* align the latter at 32 bits. So that if alignment for 64-bit values is issue, then you better arrange it in a way that makes compiler *always* align data_buffer_64bit at 64-bit boundary.

**Comment by levitte** [comment]
Ah!  I misremembered that as only applying within structures...

It just dawned on me, btw, that there's really no need for two different data buffers, as long as we start with the 64-bit values.

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: On a vaguely related **side note** one can actually wonder if this is right. I mean what's "overfeed the pool"? You don't want to overfeed the pool with entropy, right? So if you make an estimate on your source, wouldn't it be appropriate to adjust amount of data you pass down? Basically question is if "bytes_remaning" is an expression of how much entropy is missing, or is it actually how much data pool can accommodate? It's kind of ambiguous...

Comment (levitte) [reply crypto/rand/rand_vms.c]: `bytes_remaining` is the maximum available space in the pool, that's what `rand_pool_bytes_remaining()` returns.  That and the fact that `rand_pool_add()` will refuse the input entirely if it's larger than the available space in the pool should answer your question.

(on might argue that `rand_pool_add()` should accept the amount that it can, but that's outside of the scope of this PR)

File: crypto/rand/rand_vms.c
Comment (dot-asm) [inline crypto/rand/rand_vms.c]: This is no longer true. It returns number of 32-bit words.

Comment (levitte) [reply crypto/rand/rand_vms.c]: Good catch.  Fixed

**Review by dot-asm — APPROVED** [review]

**Comment by levitte** [comment]
Merged.  Thanks for the detailed review!

463e6ef500ead3fefccc470ce2f82429bb060e70 VMS: modernise rand_pool_acquire_entropy, step 2
ce147f7333fa56ab87a0878b34649991a5dae8f5 VMS: modernise rand_pool_acquire_entropy, step 1

**Pull request closed without merging**
